common-custom-user-data-gradle-plugin/src/main/java/com/gradle/BuildScanEnhancer.java

```diff
@@ -0,0 +1,442 @@
+                                        buildScan.value(test.getIdentityPath() + "#sysProps-" + key, hash(val)))
+                        )));
+    }
+
+    private void allProjects(Action<Project> action) {
+        rootProject.allprojects(action);
+    }
+
+    private static boolean hasEnv(String var) {
+        return notEmpty(System.getenv(var));
+    }
+
+    private boolean hasSystemProp(String property) {
+        return notEmpty(System.getProperty(property));
+    }
+
+    private static boolean notEmpty(String value) {
+        return value != null && value.length() > 0;
+    }
+
+    public void addCustomLinkWithSearchTerms(String title, String name, String value) {
+        // This fails with Build Scan plugin v1.16.
+        // As a workaround, you can specify the BuildScan server URL directly: e.g. `def server = "http://localhost"`
+        final String server = buildScan.getServer();
+
+        if (notEmpty(server)) {
+            String searchParams = customValueSearchParams(Map.of(name, value));
+            String url = appendIfMissing(server, "/") + "scans?" + searchParams + "#selection.buildScanB=" + urlEncode("{SCAN_ID}");
+            buildScan.link(title, url);
+        }
+
+    }
+
+    public static String customValueSearchParams(Map<String, String> search) {
+        return search.entrySet().stream()
+                .map(entry -> "search.names=" + urlEncode(entry.getKey()) + "&search.values=" + urlEncode(entry.getValue()))
+                .collect(Collectors.joining("&"));
+    }
+
+    public static String appendIfMissing(String str, String suffix) {
+        return str.endsWith(suffix) ? str : str + suffix;
+    }
+
+    public static String trimAtEnd(String str) {
+        return ("x" + str).trim().substring(1);
+    }
+
+    public static String urlEncode(String url) {
+        return URLEncoder.encode(url, StandardCharsets.UTF_8);
+    }
+
+    public static Properties readPropertiesFile(String name) {
+        final Properties properties = new Properties();
+        try (InputStream inputStream = new FileInputStream(name)) {
+            properties.load(inputStream);
+        } catch (IOException e) {
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
+        return properties;
+    }
+
+    private String hash(Object value) {
+        if (value == null) {
+            return null;
+        }
+        String string = String.valueOf(value);
+        byte[] encodedHash = messageDigest.digest(string.getBytes(StandardCharsets.UTF_8));
+        StringBuilder hexString = new StringBuilder();
+        for (int i = 0; i < encodedHash.length / 4; i++) {
+            String hex = Integer.toHexString(0xff & encodedHash[i]);
+            if (hex.length() == 1) {
+                hexString.append("0");
+            }
+            hexString.append(hex);
+        }
+        hexString.append("...");
+        return hexString.toString();
+    }
+}
```

Comment:
I suggest to copy/paste the code from [here](https://github.com/gradle/gradle-enterprise-build-config-samples/blob/master/common-custom-user-data-maven-extension/src/main/java/com/gradle/CustomBuildScanConfig.java#L28).

Reply:
Except the custom data we add only for Gradle.